packages/flutter/test/widgets/navigator_test.dart

```diff
@@ -448,6 +473,207 @@ void main() {
     expect(find.text('B'), findsOneWidget);
   });
 
+  testWidgets('pushReplacement sets secondaryAnimation after transition, with history change during transition', (WidgetTester tester) async {
+    final Map<String, SlideInOutPageRoute<dynamic>> routes = <String, SlideInOutPageRoute<dynamic>>{};
+    final Map<String, WidgetBuilder> builders = <String, WidgetBuilder>{
+      '/' : (BuildContext context) => OnTapPage(
+        id: '/',
+        onTap: () {
+          Navigator.pushNamed(context, '/A');
+        }
+      ),
+      '/A': (BuildContext context) => OnTapPage(
+        id: 'A',
+        onTap: () {
+          Navigator.pushNamed(context, '/B');
+        }
+      ),
+      '/B': (BuildContext context) => OnTapPage(
+        id: 'B',
+        onTap: () {
+          Navigator.pushReplacementNamed(context, '/C');
+        },
+      ),
+      '/C': (BuildContext context) => OnTapPage(
+        id: 'C',
+        onTap: () {
+          Navigator.removeRoute(context, routes['/']);
+        },
+      ),
+    };
+    await tester.pumpWidget(MaterialApp(
+      onGenerateRoute: (RouteSettings settings) {
+        final SlideInOutPageRoute<dynamic> ret = SlideInOutPageRoute<dynamic>(bodyBuilder: builders[settings.name], settings: settings);
+        routes[settings.name] = ret;
+        return ret;
+      }
+    ));
+    await tester.pumpAndSettle();
+    await tester.tap(find.text('/'));
+    await tester.pumpAndSettle();
+    await tester.tap(find.text('A'));
+    await tester.pumpAndSettle();
+    await tester.tap(find.text('B'));
+    await tester.pump();
+    await tester.pump(const Duration(milliseconds: 200));
```

Comment:
can you check the A's secondanimation.value equals to some expected value?

Reply:
Extra check added as requested.

As regards newIndex... it is possible that the index of the new route could change during the animation. Line ~1904 associates the preceding route with the new route at the point of pushing. Line ~1890 occurs after the animation and ensures that the new preceding route (if different) is similarly associated.
It is necessary to do the former if multiple push animations coincide, and the latter to ensure we finish in a valid state.

Reply:
I see we have different behavior on whether we notify the route before or after transition. push/replace/pop/pushReplacement notify before the transition, while pushAndRemoveUntil notify after transition. This is problematic since it might notify the wrong route. This pr change pushReplacement and pushAndRemoveUntil to notify both before and after the transition which seems wrong since it will always double notify preceding route.
I believe the intention is to only notify before the transition begin. If index change after the transition, whoever trigger the change should re-notify the route correctly.